Enable Ddoc DMD release + Ddox prerelease build of DMD docs + verbatim target for DMD

wilzbach opened the pull request:
I excluded the DDoc release build because it would be directly added to the menu (of the release build) and there's still a lot that isn't documented properly or shouldn't be exposed publicly.

However, as a quick preview here's a screenshot the Ddox menu (prerelease!):

![image](https://cloud.githubusercontent.com/assets/4370550/23199667/1599541e-f8d0-11e6-871a-d9bd8e4aa49f.png)

And ddox's awesome auto-completion:

![image](https://cloud.githubusercontent.com/assets/4370550/23199673/1e3f34bc-f8d0-11e6-8b55-73fc600f5e3e.png)

To enable dmd on the menu for the prerelease pages, https://github.com/dlang/dlang.org/pull/1580 is needed

wilzbach commented on posix.mak:
```diff
@@ -353,6 +353,11 @@ $(DMD) : ${DMD_DIR}
+dmd-release : $(STD_DDOC) $(DMD_DIR) $(DMD)
+	$(MAKE) AUTO_BOOTSTRAP=1 --directory=$(DMD_DIR) -f posix.mak html \
+		DOCDIR=${DOC_OUTPUT_DIR}/dmd \
+		DOCFMT="$(addprefix `pwd`/, $(STD_DDOC))"
```
The HTML for ddoc will get built from now on, but without a menu entry it's not visible to the user.

wilzbach commented on posix.mak:
```diff
@@ -456,13 +467,18 @@ docs.json : ${DMD_REL} ${DRUNTIME_DIR}-${LATEST} \
 	find ${PHOBOS_DIR} -name '*.d' | sed -e /unittest.d/d \
 	  -e /windows/d >> .prerelease-files.txt
-	${DMD} -c -o- -version=CoreDdoc -version=StdDdoc -Df.prerelease-dummy.html \
+	${DMD} -J$(DMD_DIR)/res -J$(DMD_DIR)/src -c -o- -version=MARS -version=CoreDdoc \
```
`MARS` is needed due to [`backend/cc.d`](https://github.com/dlang/dmd/blob/master/src/backend/cc.d#L175). It's a [default flag for all D compilers](https://github.com/dlang/dmd/blob/master/src/posix.mak#L149).

CyberShadow commented:
BTW, http://dlang.org/dmd/ currently triggers that redirect (to dmd usage docs for your platform).

CyberShadow commented:
>  Enable Ddoc DMD release + Ddoc build of DMD docs #1581 

Did you mean "DDox" instead of one of those instances of "DDoc"?

CyberShadow commented:
This seems to add some unwanted packages / top-level modules to the navigation: 

- tk
- msvc_dmc
- msvc_lib

CyberShadow commented:
Please rebase, diffs are gone :)

wilzbach commented:
> Did you mean "DDox" instead of one of those instances of "DDoc"?

Yes - sorry :/ 

> This seems to add some unwanted packages / top-level modules to the navigation:

Removed. Thanks!
As mentioned I enabled the navigation only for the DDox build.
To enable it for DDoc we need #1580.

> Please rebase, diffs are gone :)

Done :)

wilzbach commented:
> To enable it for DDoc we need #1580.

Thanks a lot for pulling this so quickly.
I rebased this PR and:
- enabled ddoc menu on the prerelease page
- added DMD output to the verbatim target

wilzbach commented on posix.mak:
```diff
@@ -55,7 +55,8 @@ STABLE_RDMD=$(STABLE_DMD_ROOT)/dmd2/$(OS)/$(if $(filter $(OS),osx),bin,bin$(MODE
 	std.windows.iunknown std.windows.registry etc.linux.memoryerror std.stream std.cstream	\
-	std.socketstream std.experimental.ndslice.internal)
+	std.socketstream std.experimental.ndslice.internal \
+	tk. msvc_dmc msvc_lib)
```
We should probably add more DMD files to this list (but this can be done after this PR)

wilzbach commented on posix.mak:
```diff
@@ -281,8 +284,8 @@ ${GENERATED}/modlist-${LATEST}.ddoc : modlist.d ${STABLE_DMD} $(DRUNTIME_STABLE_
-	$(STABLE_RDMD) modlist.d $(DRUNTIME_DIR) $(PHOBOS_DIR) $(MOD_EXCLUDES_PRERELEASE) \
-		$(addprefix --dump , object std etc core) >$@
+	$(STABLE_RDMD) modlist.d $(DRUNTIME_DIR) $(PHOBOS_DIR) $(DMD_DIR) $(MOD_EXCLUDES_PRERELEASE) \
+		$(addprefix --dump , object std etc core ddmd) >$@
```
Btw is there any logic behind this order?
Imho `core`/ Druntime is more important the the `etc` C bindings ...

CyberShadow replied:
Probably not, if anything alphabetical makes more sense than the current one, and is consistent with the sorting of the subtrees.

CyberShadow replied:
Current order is fine though.

wilzbach commented on posix.mak:
```diff
@@ -496,13 +523,18 @@ docs.json : ${DMD_REL} ${DRUNTIME_STABLE_DIR} \
+docs-prerelease.json : ${DMD} ${DMD_DIR} ${DRUNTIME_DIR} \
 		${PHOBOS_DIR_GENERATED} | dpl-docs
+	find ${DMD_DIR}/src -name '*.d' | \
+		sed -e /objc.d/d -e /mscoff/d -e '/dmd\/src\/id.d/d' \
```
The exclude of `src/id.d` was only necessary because I didn't completely clean the `dmd` cache. As I assume I won't be the only one, I added it to the "ignore list".

wilzbach commented on posix.mak:
```diff
@@ -484,6 +504,13 @@ ${DOC_OUTPUT_DIR}/library-prerelease/.htaccess : dpl_prerelease_htaccess
+	DMD_EXCLUDE += -e /scanelf/d -e /libelf/d
+else
+	DMD_EXCLUDE += -e /scanmach/d -e /libmach/d
+endif
```
DDox tries to generate the docs for all `.d` files. However for dmd this is tricky, because the `{mach, elf, mscoff}` are platform dependent. Thus the need to exclude these files (and similarly for the `objc.d` files).

CyberShadow replied:
I think these notes should be comments in the file itself.

wilzbach commented:
> BTW, http://dlang.org/dmd/ currently triggers that redirect (to dmd usage docs for your platform).

I realized that even Druntime is built under `phobos{-prerelease}` and the links are relative to the current folder only. So the easiest solution was to use `phobos{-prerelease}` as well.

CyberShadow commented:
@andralex So what's your opinion on having the `ddmd` package in the library reference's navigation tree? Is this what we want to do, and should we do this now, or wait until it's renamed to `dmd`?

wilzbach commented:
> @andralex So what's your opinion on having the ddmd package in the library reference's navigation tree? Is this what we want to do, and should we do this now, or wait until it's renamed to dmd?

I assumed that this will be an open question. Hence this PR only exposes the `ddmd` on the `prerelease` pages which aren't much known / visited anyways.
And FWIW without a menu, the (d)dmd documentation is quite useless see e.g.:
http://dlang.org/dmd-prerelease/mars.html

CC @RazvanN7 - this will be one bit that you can update soon ;-)

andralex commented:
>@andralex So what's your opinion on having the ddmd package in the library reference's navigation tree? Is this what we want to do, and should we do this now, or wait until it's renamed to dmd?

I'm not too worried about that, pages will be of narrow interest and it's unlikely anyone's life will depend on the links staying stable. Yes @RazvanN7 keep this on your list to change when we settle.

CyberShadow commented:
The filenames under `/phobos-prerelease/` don't look right.

E.g.: `web/phobos-prerelease/intrange.html`

I'm guessing should be `ddmd_intrange.html`.

This causes the navigation to be broken as well.

wilzbach commented:
> The filenames under /phobos-prerelease/ don't look right.

Well it turns out that the HTML generation script at DMD is completely different to the one at Phobos and doesn't support this `mod_ule` notation for paths.
This means we need to modify the DMD script: 

https://github.com/dlang/dmd/pull/6581

wilzbach commented:
> This means we need to modify the DMD script:
> dlang/dmd#6581

Thanks a lot for pushing this through so quickly :)

-> rebased this, so it should be ready to ship :ship:

andralex commented:
I guess I'll...

wilzbach commented:
@andralex it's now live :party:

DDoc: https://dlang.org/phobos-prerelease/ddmd_mars.html
DDox: https://dlang.org/library-prerelease/ddmd/mars.html (enjoy the search feature on the top right!)

Next steps could be:
- add more ddoc headers for public declarations to `(d)dmd`
- exclude internal modules from the list ([the list of currently excluded modules is here](https://github.com/dlang/dlang.org/blob/master/posix.mak#L73))
- prepare dlang.org for the `ddmd` -> `dmd` rename (CC @RazvanN7)
...
- flip the switch and enable it for the stable  (aka released) pages

andralex commented:
Yowzers! cc @WalterBright
